notes/testable_examples.md

```diff
@@ -0,0 +1,208 @@
+assets and thus avoid the need to duplicate them as separate files for
+distribution.
+
+## Named code blocks
+
+In the new version of odoc code blocks can be annotated with a file name. This
+file name is used by odoc to group related code blocks for extraction, and also
+to correctly annotate the markup for syntax highlighting.
+
+The following table demonstrates the two variants of code blocks: the
+traditionally supported *anonymous* code blocks and the new *named* code
+blocks.
+
+| **Anonymous code block** | **Named code block**                |
+| ------------------------ | ----------------------------------- |
+| `"{[" <content> "]}"`    | `"{" <filename> "[" <content> "]}"` |
+
+
+### Code extraction
+
+Both named and anonymous code blocks can be extracted by odoc via the
+command-line interface. Code blocks with the same file name in a given
+documentation file will be concatenated and written into a file with that name.
+Optionally a different output file name for a given group can be provided.
+Users are always required to provide an output file name for extraction of
+anonymous code blocks.
+
+To facilitate debugging and allow the tooling to implement expect-style
+promotions, popularized by cram and dune, the extracted examples can be
+optionally annotated with line numbers and the source file name (see [Line
+number directives](https://caml.inria.fr/pub/docs/manual-ocaml/lex.html#sec86)
+in the OCaml manual).
+
+**Note**: unrelated code blocks do not need to have a unique file name, it is
+recommended to group them by using a file name like `examples.ml` or similar.
+
+The described functionality will also be exposed as a library to facilitate
+integration with build systems and test promotion tooling.
+
+### Syntax highlighting
+
+The file names used to annotated code blocks are also used by odoc to decide
+what language should be used for syntax highlighting in the generated HTML. The
+language is decided based on the file name’s extension.
+
+**Warning:** code blocks without a file name will not have syntax highlighting.
+Once this feature is implemented, the currently used automatic language
+inference should be disabled.
+
+
+## Command-line interface
+
+The following simplified manual page defines the command-line interface:
+
+
+    odoc-extract-code(1)              Odoc Manual             odoc-extract-code(1)
+
+
+    NAME
+           odoc-extract-code - Extract code blocks included in documentation files.
+
+    SYNOPSIS
+           odoc extract-code [OPTION]... FILE
+
+    OPTIONS
+           --name=NAME
+               The name of the code block to extract.
+
+           -o PATH, --output=PATH
+               Output file path. If omitted, the provided NAME will be used.
+               Required for extraction of anonymous code blocks.
+
+           --anonymous
+               Extract code blocks without name. Cannot be used with the `--name'
+               option.
```

Comment:
Why not. Note however that my guess is that unless the author pays particular attention to her anonymous code block the result will likely be garbage most of the time (you may have code showing errors, or even get different languages...)

Reply:
Echoing @jsomers desire for certain fileless senarios (though you could always specificy a filename to use by convention). This could be `--anonymous=TOKEN`, and would extract anonymous code blocks that have the token `TOKEN` (e.g. `ocamlx`). 

It would also prevent the likely garbage problem I mentioned, if completely unconstrained.

Reply:
How is it _anonymous_ if it has a name (`ocamlx` in your example)?

But I do understand your concerns. I will try to address this in the new proposal for code block annotations.

Reply:
Still regarding anonymous blocks:

> The only thing I'm a little bit sceptical about is the extraction of
> anonymous blocks, it will likely often be made of garbage so I wouldn't
> bother.

It was one of the @lpw25's concerns, I believe. I imagine for large code bases
it would be useful to operate on existing unmarked code blocks in some way. So
if people do want to execute them they might get an opportunity to fix the
"garbage".

Reply:
> How is it _anonymous_ if it has a name (`ocamlx` in your example)?

In the sense it has no file*name*.

Reply:
Basically in the [two forms](https://github.com/ocaml/odoc/pull/303#discussion_r256563229) here. The first one is a named code block and the second one is an anonymous one. 

I quite like the idea of being able to extract by token, this leaves a lot of flexibity to authors to attribute meaning to what they want to extract.

Reply:
I also like the idea of token annotations. Without wanting to commit to a concrete syntax, something like `{ocaml id=BLOCK_ID[]}` could work, where `BLOCK_ID` could be anything including a file name. This of course would be used during extraction. I'd make the `--output` option required though, to clarify that the `id` is not necessarily a file name.

Reply:
I still think it's preferable to have to direct support for filenames.